File: uniquePath.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        @cache
        def helper(r, c):
            # base
            if r == m-1 and c == n-1:
                return 1
            if r == m or c == n:
                return 0

            # logic
            case1 = helper(r + 1, c)
            case2 = helper(r, c + 1)
            return case1 + case2

        
        return helper(0,0)
    
        


           
        

File: uniquepath.java
// tc: o(m*n)
// sc: o(m*n)

class Solution {
    public int uniquePaths(int m, int n) {
        int[][] mat = new int[m][n];
        
        for(int i=0;i<m;i++){
            mat[i][0]=1;
        }
        for(int i=1;i<n;i++){
            mat[0][i]=1;
        }
        
        for(int i=1;i<m;i++){
            for(int j=1;j<n;j++){
                mat[i][j]=mat[i-1][j]+mat[i][j-1];
            }
        }
        
        return mat[m-1][n-1];
    }
}
